src/protocols.c

```diff
@@ -170,19 +186,75 @@ static int read_file(const char *path) {
 
   return status;
 } /* int read_file */
+#endif
+
+#if KERNEL_SOLARIS && HAVE_KSTAT_H
+/*
+ * Retrieves all available key/value pairs for IP, ICMP, UDP and TCP from kstat
+ * modules ip, icmp, tcp and udp
+ */
+static int read_kstat(const char *mod_name) {
+  extern kstat_ctl_t *kc;
+  kstat_t *ksp_chain = NULL;
+
+  if (mod_name == NULL)
+    return -1;
+
+  if (kc == NULL)
+    return -1;
+  for (ksp_chain = kc->kc_chain; ksp_chain != NULL;
+       ksp_chain = ksp_chain->ks_next) {
+    if (strncmp(ksp_chain->ks_module, mod_name, sizeof(ksp_chain->ks_module)) ==
+            0 &&
+        ksp_chain->ks_type == KSTAT_TYPE_NAMED) {
+      kstat_named_t *kn = NULL;
+      kstat_read(kc, ksp_chain, kn);
+      kn = (kstat_named_t *)ksp_chain->ks_data;
+      for (int i = 0; (kn != NULL) && (i < ksp_chain->ks_ndata); i++, kn++) {
+        char value[16];
+        char name[256];
```

Comment:
I recommend initializing these buffers to zero, e.g.:

```c
char name[256] = "";
```

That way, if *get_kstat_value_to_string()* does not initialize the buffer due to a failure, the following code will still behave in a well-defined manner.